### PR TITLE
Better repeated MC crash handling

### PR DESCRIPTION
**Pull request by Cyberboss**
Problem subsystem will be off-lined after 3 repeated crashes.

@MrStonedOne 

Related #25808 

```
The monkey (301) scratches.
The monkey (202) scratches.
ADMIN LOG: MC crashed or runtimed, restarting
ADMIN LOG: MC crashed or runtimed, restarting
ADMIN LOG: Failed to recreate MC (Error code: 0), it's up to the failsafe now
Notice: DEFCON 3. The Master Controller has not fired in the last 40 ticks.
Warning: DEFCON 2. The Master Controller has not fired in the last 60 ticks. Automatic restart in 20 ticks.
Warning: DEFCON 1. The Master Controller has still not fired within the last 80 ticks. Killing and restarting...
The Events subsystem was the last to fire for 2 controller restarts. Will be disabled if it happens again.
MC restarted successfully
ADMIN LOG: MC crashed or runtimed, restarting
ADMIN LOG: Failed to recreate MC (Error code: 0), it's up to the failsafe now
Notice: DEFCON 3. The Master Controller has not fired in the last 40 ticks.
Warning: DEFCON 2. The Master Controller has not fired in the last 60 ticks. Automatic restart in 20 ticks.
Warning: DEFCON 1. The Master Controller has still not fired within the last 80 ticks. Killing and restarting...
The Events subsystem seems to be destabilizing the MC and will be offlined.
MC restarted successfully
The monkey (30) scratches.
Runtime stops resting.
```

### CODE REVIEW

File: code/controllers/master.dm [hunk 1]
```diff
@@ -107,6 +107,20 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
 				else
 					msg += "\t [varname] = [varval]\n"
 	log_world(msg)
+
+	var/datum/controller/subsystem/BadBoy = Master.vars["last_type_processed"]
+	if(istype(BadBoy))
+		msg = null
+		switch(++BadBoy.failure_strikes)
+			if(2)
+				msg = "The [BadBoy.name] subsystem was the last to fire after 2 controller restarts. Will be disabled if it happens again."
```
Comment (octareenroon91) [inline code/controllers/master.dm]: 'after' seems inaccurate

File: code/controllers/master.dm [hunk 2]
```diff
@@ -107,6 +107,20 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
 				else
 					msg += "\t [varname] = [varval]\n"
 	log_world(msg)
+
+	var/datum/controller/subsystem/BadBoy = Master.vars["last_type_processed"]
```
Comment (Iamgoofball) [inline code/controllers/master.dm]: what're you gonna do anyways

Comment (RemieRichards) [inline code/controllers/master.dm]: why on earth are you using the vars list.

File: code/controllers/master.dm [hunk 3]
```diff
@@ -107,6 +107,20 @@ var/CURRENT_TICKLIMIT = TICK_LIMIT_RUNNING
...
 	log_world(msg)
+
+	var/datum/controller/subsystem/BadBoy = Master.last_type_processed
+	if(istype(BadBoy))
+		msg = null
+		switch(++BadBoy.failure_strikes)
+			if(2)
+				msg = "The [BadBoy.name] subsystem was the last to fire for 2 controller restarts. Will be disabled if it happens again."
+			if(3)
+				msg = "The [BadBoy.name] subsystem seems to be destabilizing the MC and will be offlined."
```
Comment (octareenroon91) [inline code/controllers/master.dm]: Maybe I imagined it, but I thought you'd set the "stopping a subsystem" announcement to tell everyone. That choice seemed appropriate.

**Comment by AnturK** [comment]
I'm honestly not sure if we should keep going if one of subsystems burns down completely.

**Comment by Cyberboss** [comment]
That's at the admins discretion. This is just to make things slightly less
catastrophic

**Comment by JamieH** [comment]
@AnturK IMO - the MC should keep ticking. Could possibly disable the subsystem which exploded.

**Comment by Iamgoofball** [comment]
We can live without some subsystems, but we should probably mark high priority ones as critical and force a restart if they die

**Comment by octareenroon91** [comment]
>restart loop

**Comment by MrStonedOne** [comment]
before you do that, you should try to recover the subsystem by re-making and re-initializing it using the Recover() framework

subsystems aren't recreated if the mc restarts because that general causes more harm.

I'd say recreate the subsystem on the second crash, and disable on the third.

**Pull request merged**